Annotate various WTF functions with [[clang::noescape]]

rniwa opened the pull request:
#### 81fbecbe3fd4cad0f1296627466b1efa63c4f806
<pre>
Annotate various WTF functions with [[clang::noescape]]
<a href="https://bugs.webkit.org/show_bug.cgi?id=282149">https://bugs.webkit.org/show_bug.cgi?id=282149</a>

Reviewed by Geoffrey Garen.

In order to prepare WebKit for the deployment of lambda captures checker,
annotate various functions in WTF (and one function in WebCore) with
[[clang::noescape]], which is defined using NOESCAPE macro.

* Source/WTF/wtf/AggregateLogger.h:
(WTF::AggregateLogger::willLog const): Replace the use of allOf with a for
loop since clang doesn&apos;t like it.
* Source/WTF/wtf/Algorithms.h:
(WTF::forEach):
(WTF::anyOf):
(WTF::allOf):
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::createWithSizeFromGenerator):
* Source/WTF/wtf/HashMap.h:
(WTF::HashMapTranslator::translate):
(WTF::HashMapEnsureTranslator::translate):
(WTF::HashMapTranslatorAdapter::translate):
(WTF::HashMapEnsureTranslatorAdapter::translate):
(WTF::shouldValidateKey&gt;::inlineEnsure):
(WTF::shouldValidateKey&gt;::ensure):
(WTF::shouldValidateKey&gt;::removeIf):
* Source/WTF/wtf/HashSet.h:
(WTF::HashSetTranslator::translate):
(WTF::HashSetEnsureTranslatorAdaptor::translate):
(WTF::TableTraits&gt;::ensure):
(WTF::W&gt;::removeIf):
* Source/WTF/wtf/HashTable.h:
(WTF::IdentityHashTranslator::translate):
(WTF::shouldValidateKey&gt;::addUniqueForInitialization):
(WTF::shouldValidateKey&gt;::add):
(WTF::shouldValidateKey&gt;::addPassingHashCode):
(WTF::shouldValidateKey&gt;::removeIf):
(WTF::shouldValidateKey&gt;::takeIf):
* Source/WTF/wtf/LocklessBag.h:
* Source/WTF/wtf/PriorityQueue.h:
* Source/WTF/wtf/RobinHoodHashTable.h:
(WTF::SizePolicy&gt;::add):
(WTF::SizePolicy&gt;::addPassingHashCode):
* Source/WTF/wtf/StackTrace.h:
(WTF::StackTraceSymbolResolver::forEach const):
(WTF::StackTrace::forEachFrame const):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::valueOrCompute):
* Source/WTF/wtf/TinyPtrSet.h:
(WTF::TinyPtrSet::forEach const):
(WTF::TinyPtrSet::genericFilter):
* Source/WTF/wtf/TrailingArray.h:
(WTF::TrailingArray::TrailingArray):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
(WTF::Vector::containsIf const):
(WTF::Malloc&gt;::findIf const):
(WTF::Malloc&gt;::reverseFindIf const):
(WTF::Malloc&gt;::appendUsingFunctor):
(WTF::Malloc&gt;::removeFirstMatching):
(WTF::Malloc&gt;::removeLastMatching):
(WTF::Malloc&gt;::removeAllMatching):
(WTF::Malloc&gt;::map const):
(WTF:: const const):
(WTF::Malloc&gt;::appendContainerWithMapping):
(WTF::map):
(WTF::CompactMapper::compactMap):
(WTF::compactMap):
(WTF::flatMap):
* Source/WTF/wtf/WeakHashMap.h:
* Source/WTF/wtf/WeakHashSet.h:
* Source/WTF/wtf/cocoa/VectorCocoa.h:
(WTF::createNSArray):
(WTF::makeVector):
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::SignalHandlers::forEachHandler const):
* Source/WTF/wtf/threads/Signals.h:
* Source/WTF/wtf/win/SignalsWin.cpp:
(WTF::SignalHandlers::forEachHandler const):
* Source/WebCore/bindings/js/JSDOMExceptionHandling.h:
(WebCore::invokeFunctorPropagatingExceptionIfNecessary):

Canonical link: <a href="https://commits.webkit.org/285751@main">https://commits.webkit.org/285751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29df838144fd6885d1c4e4045edb8af20671ffb8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73707 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78022 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24945 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/921 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57974 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16348 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76774 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48114 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63419 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38375 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44841 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20906 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23278 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66844 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66464 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21254 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79596 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72964 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1024 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/486 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66318 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63429 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65597 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9460 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7641 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94746 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11363 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/988 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3738 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20829 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1017 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1004 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1023 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -562,6 +562,16 @@
 #define IGNORE_NULL_CHECK_WARNINGS_BEGIN IGNORE_WARNINGS_BEGIN("nonnull")
 #define IGNORE_NULL_CHECK_WARNINGS_END IGNORE_WARNINGS_END
 
+/* NOESCAPE */
```
Since the concept will be new to some people, maybe we should add a comment:

`This attribute promises that a value will only be used for the duration of the current function, and not stored to the heap or in a global for later use. You can apply it to local variables and function arguments.`

rniwa replied:
Sure.

rniwa replied:
I don't think we can apply it to a local variable though.

rniwa replied:
Added:
```
/* This attribute promises that a function argumemnt will only be used for the duration of the function,
   and not stored to the heap or in a global state for later use. The compiler does not verify this claim. */
```

geoffreygaren reviewed (approved):
r=me

rniwa committed:
Annotate various WTF functions with [[clang::noescape]]
https://bugs.webkit.org/show_bug.cgi?id=282149

Reviewed by Geoffrey Garen.

In order to prepare WebKit for the deployment of lambda captures checker,
annotate various functions in WTF (and one function in WebCore) with
[[clang::noescape]], which is defined using NOESCAPE macro.

* Source/WTF/wtf/AggregateLogger.h:
(WTF::AggregateLogger::willLog const): Replace the use of allOf with a for
loop since clang doesn't like it.
* Source/WTF/wtf/Algorithms.h:
(WTF::forEach):
(WTF::anyOf):
(WTF::allOf):
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::createWithSizeFromGenerator):
* Source/WTF/wtf/HashMap.h:
(WTF::HashMapTranslator::translate):
(WTF::HashMapEnsureTranslator::translate):
(WTF::HashMapTranslatorAdapter::translate):
(WTF::HashMapEnsureTranslatorAdapter::translate):
(WTF::shouldValidateKey>::inlineEnsure):
(WTF::shouldValidateKey>::ensure):
(WTF::shouldValidateKey>::removeIf):
* Source/WTF/wtf/HashSet.h:
(WTF::HashSetTranslator::translate):
(WTF::HashSetEnsureTranslatorAdaptor::translate):
(WTF::TableTraits>::ensure):
(WTF::W>::removeIf):
* Source/WTF/wtf/HashTable.h:
(WTF::IdentityHashTranslator::translate):
(WTF::shouldValidateKey>::addUniqueForInitialization):
(WTF::shouldValidateKey>::add):
(WTF::shouldValidateKey>::addPassingHashCode):
(WTF::shouldValidateKey>::removeIf):
(WTF::shouldValidateKey>::takeIf):
* Source/WTF/wtf/LocklessBag.h:
* Source/WTF/wtf/PriorityQueue.h:
* Source/WTF/wtf/RobinHoodHashTable.h:
(WTF::SizePolicy>::add):
(WTF::SizePolicy>::addPassingHashCode):
* Source/WTF/wtf/StackTrace.h:
(WTF::StackTraceSymbolResolver::forEach const):
(WTF::StackTrace::forEachFrame const):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::valueOrCompute):
* Source/WTF/wtf/TinyPtrSet.h:
(WTF::TinyPtrSet::forEach const):
(WTF::TinyPtrSet::genericFilter):
* Source/WTF/wtf/TrailingArray.h:
(WTF::TrailingArray::TrailingArray):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
(WTF::Vector::containsIf const):
(WTF::Malloc>::findIf const):
(WTF::Malloc>::reverseFindIf const):
(WTF::Malloc>::appendUsingFunctor):
(WTF::Malloc>::removeFirstMatching):
(WTF::Malloc>::removeLastMatching):
(WTF::Malloc>::removeAllMatching):
(WTF::Malloc>::map const):
(WTF:: const const):
(WTF::Malloc>::appendContainerWithMapping):
(WTF::map):
(WTF::CompactMapper::compactMap):
(WTF::compactMap):
(WTF::flatMap):
* Source/WTF/wtf/WeakHashMap.h:
* Source/WTF/wtf/WeakHashSet.h:
* Source/WTF/wtf/cocoa/VectorCocoa.h:
(WTF::createNSArray):
(WTF::makeVector):
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::SignalHandlers::forEachHandler const):
* Source/WTF/wtf/threads/Signals.h:
* Source/WTF/wtf/win/SignalsWin.cpp:
(WTF::SignalHandlers::forEachHandler const):
* Source/WebCore/bindings/js/JSDOMExceptionHandling.h:
(WebCore::invokeFunctorPropagatingExceptionIfNecessary):

Canonical link: https://commits.webkit.org/285751@main
